src/ziggurat/messaging/connection.clj

```diff
@@ -0,0 +1,70 @@
+(ns ziggurat.messaging.connection
+  (:require [clojure.tools.logging :as log]
+            [langohr.core :as rmq]
+            [mount.core :as mount :refer [defstate start]]
+            [sentry-clj.async :as sentry]
+            [ziggurat.config :refer [ziggurat-config]]
+            [ziggurat.sentry :refer [sentry-reporter]]
+            [ziggurat.channel :refer [get-keys-for-topic]]
+            [ziggurat.tracer :refer [tracer]]
+            [ziggurat.messaging.util :as util]
+            [clojure.string :as str])
+  (:import [com.rabbitmq.client ShutdownListener Address ListAddressResolver]
+           [java.util.concurrent Executors ExecutorService]
+           [io.opentracing.contrib.rabbitmq TracingConnectionFactory]
+           [com.rabbitmq.client.impl DefaultCredentialsProvider]))
+
+(defn is-connection-required? []
+  (let [stream-routes (:stream-routes (mount/args))
+        all-channels  (reduce (fn [all-channel-vec [topic-entity _]]
+                                (concat all-channel-vec (get-keys-for-topic stream-routes topic-entity)))
+                              []
+                              stream-routes)]
+    (or (pos? (count all-channels))
+        (-> (ziggurat-config) :retry :enabled))))
+
+(defn- channel-threads [channels]
+  (reduce (fn [sum [_ channel-config]]
+            (+ sum (:worker-count channel-config))) 0 channels))
+
+(defn- total-thread-count []
+  (let [stream-routes (:stream-router (ziggurat-config))
+        worker-count  (get-in (ziggurat-config) [:jobs :instant :worker-count])]
+    (reduce (fn [sum [_ route-config]]
+              (+ sum (channel-threads (:channels route-config)) worker-count)) 0 stream-routes)))
+
+(defn- create-traced-connection [config]
+  (let [connection-factory (TracingConnectionFactory. tracer)]
+    (.setCredentialsProvider connection-factory (DefaultCredentialsProvider. (:username config) (:password config)))
+    (.newConnection connection-factory ^ExecutorService (:executor config) ^ListAddressResolver (ListAddressResolver. (map #(Address. %) (util/list-of-hosts config))))))
+
+(defn create-connection [config tracer-enabled]
+  (if tracer-enabled
+    (create-traced-connection config)
+    (rmq/connect (assoc config :hosts (util/list-of-hosts config)))))
```

Comment:
this will break backward compatibility for services using a single host rabbitMQ